modules: chre

yperess opened the pull request:
Add initial build rules for CHRE. This change includes a Kconfig and
CMakeLists.txt to begin compiling code from the CHRE module.
Additional files are included to bridge the APIs from CHRE's to
Zephyr's. These can be found in modules/chre/include and
modules/chre/src.

Additionally, add a sample to make sure that the module builds. It can
be built via:

```
$ west build -b native_posix -p=always samples/application_development/chre
```

Issue #37223
Signed-off-by: Yuval Peress <peress@google.com>

Commit:
modules: chre

Add initial build rules for CHRE. This change includes a Kconfig and
CMakeLists.txt to begin compiling code from the CHRE module.
Additional files are included to bridge the APIs from CHRE's to
Zephyr's. These can be found in modules/chre/include and
modules/chre/src.

Additionally, add a sample to make sure that the module builds. It can
be built via:

```
$ west build -b native_posix -p=always samples/application_development/chre
```

Signed-off-by: Yuval Peress <peress@google.com>

tejlmand commented on west.yml:
```diff
@@ -28,6 +28,10 @@ manifest:
       revision: 1052dae561497bef901f931ef75e117c9224aecd
       path: modules/lib/canopennode
+    - name: chre
+      url: https://android.googlesource.com/platform/system/chre
```
if module is approved we'll need a Zephyr fork of this.

yperess replied:
Done

tejlmand commented on modules/chre/Kconfig:
```diff
@@ -0,0 +1,54 @@
+menuconfig CHRE
+	bool "CHRE Support"
+	help
+	  This option enables the CHRE library.
```
```suggestion
	  This option enables the Context Hub Runtime Environment (CHRE) library.
```

tejlmand commented on samples/application_development/chre/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(chre_sample)
```
Maybe write name in full, or as minimum a comment here with full name of what chre is.

yperess replied:
I'll be adding a full doc to this directory explaining everything

tejlmand commented on samples/application_development/chre/src/main.c:
```diff
@@ -0,0 +1,11 @@
+#include <zephyr.h>
+#include <sys/printk.h>
+
+void main(void)
+{
+	printk("Hello CHRE!\n");
```
In which way is this sample using / exemplifying how chre can be used in Zephyr context ?

MaureenHelm replied:
This sample needs to use CHRE in some way, and it should be moved to `samples/modules/chre`.

Also, the sample doesn't build for me. CI didn't catch it because the sample is missing a `sample.yaml`
```
/home/mhelm/zephyrproject/modules/lib/chre/platform/shared/version.cc:29:19: error: expected unqualified-id before string constant
   29 | DLL_EXPORT extern "C" const char _chreVersionString[]
      |                   ^~~
/home/mhelm/zephyrproject/modules/lib/chre/platform/shared/version.cc: In function ‘const char* chre::getChreVersionString()’:
/home/mhelm/zephyrproject/modules/lib/chre/platform/shared/version.cc:34:10: error: ‘_chreVersionString’ was not declared in this scope; did you mean ‘getChreVersionString’?
   34 |   return _chreVersionString;
      |          ^~~~~~~~~~~~~~~~~~
      |          getChreVersionString
```

yperess replied:
Done

tejlmand reviewed (changes requested):
Thanks for this contribution.

A couple of blockers that must be addressed:
- In no places it is described what CHRE means.
- No description of what chre does, or how to use it with Zephyr.
- The sample here: https://github.com/zephyrproject-rtos/zephyr/blob/188a7d0f7d6af350db48ee7fe83d5da94b19515f/samples/application_development/chre/CMakeLists.txt lacks documentation, please add a `README.txt`
- A Zephyr fork of [chre](https://github.com/zephyrproject-rtos/zephyr/blob/188a7d0f7d6af350db48ee7fe83d5da94b19515f/west.yml#L32) must be created when rest of PR is ready for merge.

MaureenHelm commented on modules/chre/CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+  )
+
+  # Optional audio support
+  if(DEFINED CHRE_AUDIO_SUPPORT_ENABLED)
```
I think there are a few places where you need to prefix the Kconfig symbols with "CONFIG_"
```suggestion
if(CONFIG_CHRE_AUDIO_SUPPORT_ENABLED)
```

yperess replied:
Thanks, fixed

yperess replied:
Done

yperess commented:
OK, I've update the example in this PR. I'm working with the CHRE team to see if they're OK owning the build files. I think that will be the best approach. The new sample actually runs the full CHRE, loads a nanoapp, sends an event, etc.

The goal is to have everything under `modules/chre/include` and `modules/chre/src` actually live upstream so please ignore any style issues

yperess commented:
Marking as WIP as I'm moving the build files over to AOSP. When done this PR will only be the added module and the sample app. At that point I'll remove the WIP label.

henrikbrixandersen commented on modules/chre/Kconfig:
```diff
@@ -0,0 +1,70 @@
+	  When enabled, CHRE will include various assertions which will trigger
+	  calls to chreAbort if resolve to false.
+
+config CHRE_AUDIO_SUPPORT_ENABLED
```
```suggestion
config CHRE_AUDIO_SUPPORT
```

I would prefer to remove the `_ENABLED` postfix for all these Kconfig options.

yperess replied:
Please comment on https://android-review.googlesource.com/c/platform/system/chre/+/1952616

yperess commented:
PR updated now that AOSP is willing to own all the build rules for Zephyr as well as adding this to their CI. To verify this PR please also cherry-pick the 2 changes at https://android-review.googlesource.com/c/platform/system/chre/+/1952616

yperess commented:
I'd like to remind anyone interested in this PR to please review the change in https://android-review.googlesource.com/c/platform/system/chre/+/1952616/

MaureenHelm commented on samples/application_development/chre/README.rst:
```diff
@@ -0,0 +1,47 @@
+#. Update the ``overlay.dts`` of this sample application to include 2 emulated accelerometers and
+   configure them to return scripted data.
+#. Run this sample application and watch the nanoapp provide lid angle calculations based on 2
+   accelerometers provided by the sensors PAL framework.
```
Please add to the sample application README some instructions on how to build and a description of the expected output.

yperess replied:
Done

MaureenHelm reviewed (changes requested):
I'd still like to see the sample moved to `samples/modules/chre`. 
https://github.com/zephyrproject-rtos/zephyr/pull/41735#discussion_r784308374

yperess commented on samples/application_development/chre/sample.yaml:
```diff
@@ -0,0 +1,21 @@
+      - "Event \\(1\\) complete!"
+      - "EchoApp::nanoappEnd\\(\\)"
+      - "\\[.*\\] .*<inf> chre: Exiting EventLoop.*"
+    platform_allow: native_posix
```
Done

yperess replied:
Done

yperess commented on samples/application_development/chre/src/main.cc:
```diff
@@ -0,0 +1,47 @@
+
+	printk("Hello CHRE!\n");
+
+	k_sleep(K_MSEC(500));
```
Done

MaureenHelm reviewed (changes requested):
Looking good, thanks for the changes.

I tried the sample on `frdm_k64f` and got some build warnings:
```
$ west build -b frdm_k64f samples/modules/chre/ -- -DCONFIG_NEWLIB_LIBC=y
...<snip>...
[131/176] Building CXX object modules/chre/CMakeFiles/..__modules__li...hyr.dir/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc.obj
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc: In member function 'void chre::Nanoapp::logStateToBuffer(chre::DebugDumpWrapper&) const':
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'long unsigned int' [-Wformat=]
  158 |   debugDump.print(" v%" PRIu32 ".%" PRIu32 ".%" PRIu32 " tgtAPI=%" PRIu32
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  159 |                   ".%" PRIu32 " curAlloc=%zu peakAlloc=%zu",
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'long unsigned int' [-Wformat=]
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 5 has type 'long unsigned int' [-Wformat=]
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 6 has type 'long unsigned int' [-Wformat=]
/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 7 has type 'long unsigned int' [-Wformat=]
[166/176] Linking CXX executable zephyr/zephyr_pre0.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'

[170/176] Linking CXX executable zephyr/zephyr_pre1.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'

[176/176] Linking CXX executable zephyr/zephyr.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'
Memory region         Used Size  Region Size  %age Used
           FLASH:       33726 B         1 MB      3.22%
            SRAM:       17088 B       192 KB      8.69%
        IDT_LIST:          0 GB         2 KB      0.00%
```

But it runs successfully:
```
*** Booting Zephyr OS build zephyr-v3.0.0-824-g32fc27bfc09a  ***
Failed to set thread name to "CHRE": -88
Hello CHRE!
[00:00:00.010,000] <inf> chre: CHRE init, version: undefined
[00:00:00.010,000] <inf> chre: EventLoop start
EchoApp::nanoappStart()
Starting EchoApp... SUCCESS
Nanoapp count=1
Finding instance ID... SUCCESS
Nanoapp count=1
Instance ID: 1
Sending event 1...
EchoApp::nanoappHandleEvent(sender_instance_id=0, event_type=1, event_data@(nil))
Event (1) complete!
Ending EchoApp... FAIL
Goodbye!
EchoApp::nanoappEnd()
[00:00:00.510,000] <dbg> chre: startNanoapp: Instance ID 1 assigned to app ID 0x0000000000000001
[00:00:01.024,000] <err> chre: Refusing to unload system nanoapp
[00:00:01.028,000] <inf> chre: Exiting EventLoop
[00:00:01.028,000] <dbg> chre: deinit: CHRE deinit
```

yperess commented:
Current presubmit errors seem unrelated:

```
devicetree error: /__w/zephyr/zephyr/dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83 (column 18): parse error: expected number or parenthesized expression
```

yperess commented on samples/modules/chre/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+To build the sample use the following west command::
+
+  $ west build -b ${TARGET_PLATFORM} -p=always samples/modules/chre
```
Merged

yperess commented on samples/modules/chre/sample.yaml:
```diff
@@ -0,0 +1,23 @@
+    Environment (CHRE).
+  name: chre
+tests:
+  sample.application_development.chre:
```
Merged

yperess commented:
> Looking good, thanks for the changes.
> 
> I tried the sample on `frdm_k64f` and got some build warnings:
> 
> ```
> $ west build -b frdm_k64f samples/modules/chre/ -- -DCONFIG_NEWLIB_LIBC=y
> ...<snip>...
> [131/176] Building CXX object modules/chre/CMakeFiles/..__modules__li...hyr.dir/home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc.obj
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc: In member function 'void chre::Nanoapp::logStateToBuffer(chre::DebugDumpWrapper&) const':
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'long unsigned int' [-Wformat=]
>   158 |   debugDump.print(" v%" PRIu32 ".%" PRIu32 ".%" PRIu32 " tgtAPI=%" PRIu32
>       |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   159 |                   ".%" PRIu32 " curAlloc=%zu peakAlloc=%zu",
>       |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'long unsigned int' [-Wformat=]
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 5 has type 'long unsigned int' [-Wformat=]
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 6 has type 'long unsigned int' [-Wformat=]
> /home/mhelm/zephyrproject/modules/lib/chre/core/nanoapp.cc:158:19: warning: format '%u' expects argument of type 'unsigned int', but argument 7 has type 'long unsigned int' [-Wformat=]
> [166/176] Linking CXX executable zephyr/zephyr_pre0.elf
> /home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'
> 
> [170/176] Linking CXX executable zephyr/zephyr_pre1.elf
> /home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'
> 
> [176/176] Linking CXX executable zephyr/zephyr.elf
> /home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'
> Memory region         Used Size  Region Size  %age Used
>            FLASH:       33726 B         1 MB      3.22%
>             SRAM:       17088 B       192 KB      8.69%
>         IDT_LIST:          0 GB         2 KB      0.00%
> ```
> 
> But it runs successfully:
> 
> ```
> *** Booting Zephyr OS build zephyr-v3.0.0-824-g32fc27bfc09a  ***
> Failed to set thread name to "CHRE": -88
> Hello CHRE!
> [00:00:00.010,000] <inf> chre: CHRE init, version: undefined
> [00:00:00.010,000] <inf> chre: EventLoop start
> EchoApp::nanoappStart()
> Starting EchoApp... SUCCESS
> Nanoapp count=1
> Finding instance ID... SUCCESS
> Nanoapp count=1
> Instance ID: 1
> Sending event 1...
> EchoApp::nanoappHandleEvent(sender_instance_id=0, event_type=1, event_data@(nil))
> Event (1) complete!
> Ending EchoApp... FAIL
> Goodbye!
> EchoApp::nanoappEnd()
> [00:00:00.510,000] <dbg> chre: startNanoapp: Instance ID 1 assigned to app ID 0x0000000000000001
> [00:00:01.024,000] <err> chre: Refusing to unload system nanoapp
> [00:00:01.028,000] <inf> chre: Exiting EventLoop
> [00:00:01.028,000] <dbg> chre: deinit: CHRE deinit
> ```

I'll see if I can fix this upstream

MaureenHelm commented:
> Current presubmit errors seem unrelated:
> 
> ```
> devicetree error: /__w/zephyr/zephyr/dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83 (column 18): parse error: expected number or parenthesized expression
> ```

Please rebase to pick the fix in #43678

MaureenHelm commented:
> > Current presubmit errors seem unrelated:
> > ```
> > devicetree error: /__w/zephyr/zephyr/dts/arm64/nxp/nxp_imx8mm_a53.dtsi:83 (column 18): parse error: expected number or parenthesized expression
> > ```
> 
> Please rebase to pick the fix in #43678

nvm. @nashif took care of it

nashif commented:
> nvm. @nashif took care of it

we now have the magic rebase button :-D

nashif commented:
```
/home/runner/work/zephyr/zephyr/doc/_build/src/samples/modules/chre/README.rst:33:Unexpected indentation.
FAILED: CMakeFiles/html 
```

yperess commented:
> ```
> /home/runner/work/zephyr/zephyr/doc/_build/src/samples/modules/chre/README.rst:33:Unexpected indentation.
> FAILED: CMakeFiles/html 
> ```

Fixed now

yperess commented:
@MaureenHelm @nashif I believe that the issue with the printf formatting boils down to https://lists.zephyrproject.org/g/devel/topic/uint32_t_typedef_differences/19721373. Though reading the thread I'm not really seeing a way around it and I hope that Anas might be able to shed some light about this discussion. The code in CHRE is using `UINT32_C()` for the value and `PRIu32` for the format string. This seems like it should work, but I'm understanding it has an issue with newlibc?

tejlmand reviewed (dismissed):
why are the files named `.cc` and `.hh` and not just `.cpp` and `.hpp`.

yperess commented:
> why are the files named `.cc` and `.hh` and not just `.cpp` and `.hpp`.

There's no hard reason for it. The C++ libraries I've integrated with Zephyr so far (CHRE and Pigweed) both use `.cc`. The header extension `.hh` was selected to follow the same pattern and was needed to avoid the linter thinking that the header should be formatter for C (this was an issue with templates as clang-format thought there needed to be spaces around the `<`). I don't believe that Zephyr has a specific guideline for this yet do they? I also believe that the Unix standard is to use `.cc` which most of the Zephyr guides model after so that's why I chose it. I'm happy to change if anyone feels strongly one way or another.

MaureenHelm commented:
> @MaureenHelm @nashif I believe that the issue with the printf formatting boils down to [lists.zephyrproject.org/g/devel/topic/uint32_t_typedef_differences/19721373](https://lists.zephyrproject.org/g/devel/topic/uint32_t_typedef_differences/19721373). Though reading the thread I'm not really seeing a way around it and I hope that Anas might be able to shed some light about this discussion. The code in CHRE is using `UINT32_C()` for the value and `PRIu32` for the format string. This seems like it should work, but I'm understanding it has an issue with newlibc?

We need to find a solution, because even though CI passes on this PR, the more extensive nightly CI run will fail. The native_posix target is handy for a lot of things, but we need to make sure this works on at least one embedded target.

yperess commented:
> > @MaureenHelm @nashif I believe that the issue with the printf formatting boils down to [lists.zephyrproject.org/g/devel/topic/uint32_t_typedef_differences/19721373](https://lists.zephyrproject.org/g/devel/topic/uint32_t_typedef_differences/19721373). Though reading the thread I'm not really seeing a way around it and I hope that Anas might be able to shed some light about this discussion. The code in CHRE is using `UINT32_C()` for the value and `PRIu32` for the format string. This seems like it should work, but I'm understanding it has an issue with newlibc?
> 
> We need to find a solution, because even though CI passes on this PR, the more extensive nightly CI run will fail. The native_posix target is handy for a lot of things, but we need to make sure this works on at least one embedded target.

Thanks @MaureenHelm, I believe that https://android-review.googlesource.com/c/platform/system/chre/+/2030923 should do the trick.

tejlmand commented:
> There's no hard reason for it. The C++ libraries I've integrated with Zephyr so far (CHRE and Pigweed) both use `.cc`. The header extension `.hh` was selected to follow the same pattern and was needed to avoid the linter thinking that the header should be formatter for C

Thanks for the reasoning.
I think we need to find out what pattern Zephyr should use.
Cause other modules, such as [openthread](https://github.com/zephyrproject-rtos/openthread/tree/zephyr/src/core/radio), [mipi-sys-t](https://github.com/zephyrproject-rtos/mipi-sys-t/tree/zephyr/printer/src) and more uses `cpp` / `hpp` and within Zephyr `cpp` is used in a very few places: [zephyr](https://github.com/zephyrproject-rtos/zephyr/tree/main/subsys/cpp)

Of course Zephyr shouldn't be dictated to a specific form because of external repos, but personally I would like to see a guideline on which form Zephyr would prefer.
Multiple forms inside same repo is just creating confusion.

@nashif @galak @carlescufi @MaureenHelm @stephanosio I don't see anything in our guidelines regarding this, but I think it's any important topic to consider.
https://docs.zephyrproject.org/latest/contribute/index.html#

yperess commented:
@tejlmand is there a particular meeting next week that we should add this to the agenda?

carlescufi commented:
> @tejlmand is there a particular meeting next week that we should add this to the agenda?

Unfortunatel @tejlmand is currently on parental leave. I suggest you raise this in today's dev-review or next week's API meeting.

stephanosio commented:
> @tejlmand is there a particular meeting next week that we should add this to the agenda?

I have created the issue #44196 to keep track of this, and flagged it for the next dev-review.

MaureenHelm commented on include/arch/structs.h:
```diff
@@ -31,6 +31,12 @@
 
 /* Per CPU architecture specifics (empty) */
 struct _cpu_arch {
+#ifdef __cplusplus
+	/* This struct will have a size 0 in C which is not allowed in C++ (it'll have a size 1). To
+	 * prevent this, we add a 1 byte dummy variable.
+	 */
+	uint8_t dummy;
+#endif
```
Please pull the changes in `include/` into a separate commit

yperess replied:
Done

MaureenHelm reviewed (changes requested):
I got a new linker warning when building for `frdm_k64f`:

```
$ west build -p auto -b frdm_k64f samples/modules/chre/ -- -DCONFIG_NEWLIB_LIBC=y
<snip>
[164/174] Linking CXX executable zephyr/zephyr_pre0.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'

[168/174] Linking CXX executable zephyr/zephyr_pre1.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'

[174/174] Linking CXX executable zephyr/zephyr.elf
/home/mhelm/.local/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `.unstable_id' from `modules/chre/lib..__modules__lib__chre__platform__zephyr.a(version.cc.obj)' being placed in section `.unstable_id'
Memory region         Used Size  Region Size  %age Used
           FLASH:       33454 B         1 MB      3.19%
            SRAM:       17088 B       192 KB      8.69%
           SRAML:          0 GB        64 KB      0.00%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Commit:
cpp: Update structs to be compatible with C++

Both the `_cpu_arch` and `k_thread_runtime_stats` structs can have a
size of 0 in C, but will fail when building with C++. Add an extra
byte in those cases.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
@MaureenHelm @nashif the upstream change landed in AOSP for supporting the linker section, but I seem to have lost permissions to https://github.com/zephyrproject-rtos/chre so I can't update the module.

stephanosio commented on west.yml:
```diff
@@ -28,6 +28,9 @@ manifest:
       revision: 53d3415c14d60f8f4bfca54bfbc5d5a667d7e724
       path: modules/lib/canopennode
+    - name: chre
+      revision: 2e1ba542aeea185d801830490e20fc93f436912f
```
> Need to use https://github.com/zephyrproject-rtos/chre/pull/1 to get rid of the warnings

```suggestion
      revision: pull/1/head
```

MaureenHelm replied:
This change indeed eliminates the warnings

yperess commented:
@MaureenHelm I update the module commit hash from the `pull`

cfriedt reviewed (approved):
Nice! LGTM👍
